src/Files.App/Filesystem/WindowsStorageDeviceWatcher.cs

```diff
@@ -0,0 +1,140 @@
+﻿using Files.App.DataModels.NavigationControlItems;
+using Files.App.Helpers;
+using Files.App.Helpers.MMI;
+using Files.Backend.Models;
+using Files.Sdk.Storage.LocatableStorage;
+using Microsoft.Extensions.Logging;
+using System;
+using System.IO;
+using Windows.Devices.Enumeration;
+using Windows.Devices.Portable;
+using Windows.Storage;
+
+namespace Files.App.Filesystem
+{
+	public class WindowsStorageDeviceWatcher : IStorageDeviceWatcher
+	{
+		public event EventHandler<ILocatableFolder> DeviceAdded;
+		public event EventHandler<string> DeviceRemoved;
+		public event EventHandler EnumerationCompleted;
+		public event EventHandler<string> DeviceModified;
+
+		private DeviceWatcher watcher;
+
+		public bool CanBeStarted => watcher.Status is DeviceWatcherStatus.Created or DeviceWatcherStatus.Stopped or DeviceWatcherStatus.Aborted;
+
+		public WindowsStorageDeviceWatcher()
+		{
+			watcher = DeviceInformation.CreateWatcher(StorageDevice.GetDeviceSelector());
+			watcher.Added += Watcher_Added;
+			watcher.Removed += Watcher_Removed;
+			watcher.EnumerationCompleted += Watcher_EnumerationCompleted;
+
+			SetupWin32Watcher();
+		}
+
+		private void SetupWin32Watcher()
+		{
+			DeviceManager.Default.DeviceAdded += Win32_OnDeviceAdded;
+			DeviceManager.Default.DeviceRemoved += Win32_OnDeviceRemoved;
+			DeviceManager.Default.DeviceInserted += Win32_OnDeviceEjectedOrInserted;
+			DeviceManager.Default.DeviceEjected += Win32_OnDeviceEjectedOrInserted;
+		}
+
+		private void Win32_OnDeviceEjectedOrInserted(object? sender, DeviceEventArgs e)
+		{
+			DeviceModified?.Invoke(this, e.DeviceId);
+		}
+
+		private void Win32_OnDeviceRemoved(object? sender, DeviceEventArgs e)
+		{
+			DeviceRemoved?.Invoke(this, e.DeviceId);
+		}
+
+		private async void Win32_OnDeviceAdded(object? sender, DeviceEventArgs e)
+		{
+			var driveAdded = new DriveInfo(e.DeviceId);
+			var rootAdded = await FilesystemTasks.Wrap(() => StorageFolder.GetFolderFromPathAsync(e.DeviceId).AsTask());
+			if (!rootAdded)
+			{
+				App.Logger.LogWarning($"{rootAdded.ErrorCode}: Attempting to add the device, {e.DeviceId},"
+					+ " failed at the StorageFolder initialization step. This device will be ignored.");
+				return;
+			}
+
+			DriveItem driveItem;
+			using (var thumbnail = await DriveHelpers.GetThumbnailAsync(rootAdded.Result))
+			{
+				var type = DriveHelpers.GetDriveType(driveAdded);
+				driveItem = await DriveItem.CreateFromPropertiesAsync(rootAdded, e.DeviceId, type, thumbnail);
```

Comment:
Maybe instead of retrieving the thumbnail, modify the method so that the thumbnail is optional? (Since we don't need it there). Ideally, you'd only notify that a device was added by providing an id (or ILocatableFolder) to it. Other classes would subscribe to this event and retrieve additional properties as needed.